R/derive_param_extreme_record.R

```diff
@@ -0,0 +1,221 @@
+                                        order,
+                                        mode,
+                                        set_values_to) {
+  # Check arguments assertions
+  assert_data_frame(dataset, optional = TRUE)
+  assert_list_of(sources, "records_source")
+  assert_list_of(source_datasets, "data.frame")
+  assert_vars(by_vars, optional = TRUE)
+  assert_character_scalar(
+    mode,
+    values = c("first", "last"),
+    case_sensitive = FALSE
+  )
+  assert_varval_list(set_values_to, accept_expr = TRUE, optional = TRUE)
+
+  source_names <- names(source_datasets)
+
+  # Create Empty list to contain source datasets
+  data_list <- vector("list", length(sources))
+
+  # Evaluate the expressions contained in the sources
+  for (i in seq_along(sources)) {
+    source_dataset <- source_datasets[[sources[[i]]$dataset_name]]
+    new_vars_colnames <- replace_values_by_names(sources[[i]]$new_vars)
+    data_list[[i]] <- source_dataset %>%
+      filter_if(sources[[i]]$filter) %>%
+      mutate(!!!sources[[i]]$new_vars) %>%
+      select(!!!by_vars, !!!new_vars_colnames)
+  }
+
+  # Bind the source datasets together and parse out the extreme value
+  param_data <- bind_rows(data_list) %>%
+    filter_extreme(.,
+      by_vars = by_vars,
+      order = order,
+      mode = mode
+    ) %>%
+    process_set_values_to(
+      set_values_to
+    )
+
+  # Bind the parameter rows back to original dataset
+  bind_rows(dataset, param_data)
+}
+
+#' Create a `records_source` Object
+#'
+#' The `records_source` object is used to find extreme records of interest.
+#'
+#' @param dataset_name The name of the source dataset
+#'
+#'   The name refers to the dataset provided by the `source_datasets` argument
+#'   of `derive_param_extreme_record()`.
+#'
+#' @param filter An unquoted condition for selecting the observations from
+#'   `dataset`.
+#'
+#' @param new_vars Variables to add
+#'
+#'   The specified variables from the source datasets are added to the output
+#'   dataset. Variables can be renamed by naming the element, i.e., `new_vars =
+#'   exprs(<new name> = <old name>)`.
+#'
+#'   For example `new_vars = exprs(var1, var2)` adds variables `var1` and `var2`
+#'   from to the input dataset.
+#'
+#'   And `new_vars = exprs(var1, new_var2 = old_var2)` takes `var1` and
+#'   `old_var2` from the source dataset and adds them to the input dataset renaming
+#'   `old_var2` to `new_var2`. Expressions can be used to create new variables
+#'   (see for example `new_vars` argument in `derive_vars_merged()`).
+#'
+#'   *Permitted Values:* list of expressions created by `exprs()`, e.g.,
+#'   `exprs(ADT, desc(AVAL))`
+#'
+#' @keywords source_specifications
+#' @family source_specifications
+#'
+#' @seealso [derive_param_extreme_record()]
+#'
+#' @return  An object of class `records_source`
+#' @export
+records_source <- function(dataset_name,
+                           filter = NULL,
+                           new_vars = NULL) {
```

Comment:
I would make `new_vars` mandatory. At the moment no variables are added if `new_vars` is not specified. But this is not useful.

Reply:
Updated!